File: wagerproof-ios-native/Wagerproof/Features/MLB/Components/MLBTeamLogo.swift
import SwiftUI
import WagerproofDesign
import WagerproofModels

/// Team logo on a team-tinted disc, with an abbreviation fallback. Mirrors
/// RN `TeamLogo` in MLBGameCard.tsx. Used by the MLB bottom sheet, the
/// betting trends sheet, and the trends situation section — extracted out
/// of `MLBGameCard.swift` when the home feed switched to the shared
/// `GameRowCard` (which handles avatar rendering itself).
struct MLBTeamLogo: View {
    let logoUrl: String?
    let abbrev: String
    let name: String
    var size: CGFloat = 42

    @Environment(\.colorScheme) private var colorScheme

    var body: some View {
        let pair = MLBTeams.colors(for: name.isEmpty ? abbrev : name)
        let primary = Color(hex: Int(pair.primary))
        // Disc is a neutral surface washed with the team's PRIMARY only. A
        // primary→secondary gradient filled half the disc with the team's
        // secondary (STL navy behind a red logo), which reads as another
        // team's background sitting behind the logo.
        let plate = logoContrastPlate(for: primary)

        ZStack {
            Circle()
                .fill(Color.appSurfaceElevated)
                .overlay(Circle().fill(primary).opacity(colorScheme == .dark ? 0.30 : 0.22))
                .overlay(Circle().strokeBorder(primary, lineWidth: 2))

            if let urlString = logoUrl, let url = URL(string: urlString) {
                CachedAsyncImage(url: url) { phase in
                    switch phase {
                    case .success(let image):
                        ZStack {
                            if let plate {
                                Circle().fill(plate).padding(size * 0.08)
                            }
                            image.resizable().scaledToFit()
                        }
                    default:
                        Text(abbrev)
                            .font(.system(size: size * 0.32, weight: .bold))
                            .foregroundStyle(Color.appTextPrimary)
                    }
                }
                .frame(width: size, height: size)
                .clipShape(Circle())
            } else {
                Text(abbrev)
                    .font(.system(size: size * 0.32, weight: .bold))
                    .foregroundStyle(Color.appTextPrimary)
            }
        }
        .frame(width: size, height: size)
    }

    /// Plate to separate a same-color logo from the disc. Only needed in light
    /// mode: a light team primary on the light surface washes the logo out. In
    /// dark mode the disc is already dark, and a lightening plate there is what
    /// produced the muddy off-team blob behind the logo.
    private func logoContrastPlate(for primary: Color) -> Color? {
        guard colorScheme != .dark else { return nil }
        return primary.relativeLuminance > 0.6 ? Color.black.opacity(0.55) : nil
    }
}

/// Shared MLB-specific formatting helpers. Mirrors `types/mlb.ts`. Used by
/// the bottom sheet, betting trends sheet, and the row card; extracted out
/// of `MLBGameCard.swift` so the trimmed home-feed wrapper can stay
/// focused on adapting to `GameRowCard`.
enum MLBFormatting {
    static func moneyline(_ ml: Int?) -> String {
        guard let ml else { return "-" }
        return ml > 0 ? "+\(ml)" : "\(ml)"
    }

    static func spread(_ s: Double?) -> String {
        guard let s, !s.isNaN else { return "-" }
        let body: String = (s == s.rounded()) ? "\(Int(s))" : String(format: "%.1f", s)
        return s > 0 ? "+\(body)" : body
    }

    static func line(_ value: Double?) -> String {
        guard let value else { return "-" }
        if value == value.rounded() { return "\(Int(value))" }
        return String(format: "%.1f", value)
    }

    // Formatters are hoisted to statics because these helpers are called from
    // the detail hero's `topRow`, which SwiftUI re-evaluates on every scroll
    // frame. Constructing a DateFormatter is expensive (locale + timezone +
    // format parsing), and this was 4-5 of them per frame at 120Hz.
    private static let etTimeZone = TimeZone(identifier: "America/New_York")

    private static let dateParser: DateFormatter = {
        let f = DateFormatter()
        f.locale = Locale(identifier: "en_US_POSIX")
        f.timeZone = etTimeZone
        f.dateFormat = "yyyy-MM-dd"
        return f
    }()

    private static let dateLabelFormatter: DateFormatter = {
        let f = DateFormatter()
        f.locale = Locale(identifier: "en_US")
        f.timeZone = etTimeZone
        f.dateFormat = "EEE, MMM d"
        return f
    }()

    private static let timeFormatter: DateFormatter = {
        let f = DateFormatter()
        f.locale = Locale(identifier: "en_US")
        f.timeZone = etTimeZone
        f.dateFormat = "h:mm a"
        return f
    }()

    /// Two ISO parsers rather than one whose `formatOptions` get mutated —
    /// a shared static that rewrites its own options is not thread-safe.
    private static let isoFractional: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return f
    }()

    private static let isoPlain: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime]
        return f
    }()

    /// Mirrors RN `formatMLBDateLabel`: short weekday + month + day from
    /// `YYYY-MM-DD`.
    static func dateLabel(_ raw: String) -> String {
        guard let date = dateParser.date(from: raw) else { return raw }
        return dateLabelFormatter.string(from: date)
    }

    /// Mirrors RN `formatMLBGameTime` — parses ISO UTC, formats `h:mm a ET`.
    static func gameTime(_ raw: String?) -> String {
        guard let raw, !raw.isEmpty else { return "TBD" }
        guard let date = isoFractional.date(from: raw) ?? isoPlain.date(from: raw) else { return "TBD" }
        return timeFormatter.string(from: date) + " ET"
    }
}

File: wagerproof-android-native/app/src/main/java/com/wagerproof/app/features/mlb/MLBTeamLogo.kt
package com.wagerproof.app.features.mlb

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.wagerproof.app.features.gamecards.TeamColorPair
import com.wagerproof.app.features.gamecards.relativeLuminance
import com.wagerproof.app.features.shared.RemoteImage
import com.wagerproof.app.features.shared.hexColor
import com.wagerproof.core.models.MLBTeams
import java.time.ZoneId
import java.time.ZonedDateTime
import java.time.format.DateTimeFormatter
import java.util.Locale

/**
 * MLB team avatar — port of iOS `MLBTeamLogo`. A gradient team-color disc with a
 * 2dp primary border, the team logo (or abbr initials) on top, plus the same
 * faint contrast plate the feed avatars use so a same-color logo separates from
 * the disc. Used by the MLB detail sheet, betting-trends card, and projections.
 */
@Composable
fun MLBTeamLogo(
    logoUrl: String?,
    abbrev: String,
    name: String,
    size: Dp,
    modifier: Modifier = Modifier,
) {
    val pair = mlbTeamColorPair(name.ifEmpty { abbrev })
    val primary = pair.primary
    val secondary = pair.secondary
    // Dark theme only (app is dark) — plate a dark logo with a faint light wash.
    val plate: Color? = if (primary.relativeLuminance() < 0.45f) Color.White.copy(alpha = 0.15f) else null
    Box(
        modifier
            .size(size)
            .clip(CircleShape)
            .background(Brush.linearGradient(listOf(primary, secondary)))
            .border(2.dp, primary, CircleShape),
        contentAlignment = Alignment.Center,
    ) {
        RemoteImage(
            url = logoUrl,
            contentDescription = abbrev,
            modifier = Modifier.size(size).padding(size * 0.08f),
            error = {
                Text(
                    abbrev,
                    color = contrastingText(primary),
                    fontSize = (size.value * 0.32f).sp,
                    fontWeight = FontWeight.Bold,
                )
            },
        )
        if (plate != null && !logoUrl.isNullOrBlank()) {
            // Faint plate lives behind the logo image; RemoteImage draws over it.
            Box(Modifier.size(size * 0.84f).clip(CircleShape).background(plate))
        }
    }
}

/** Luminance-contrast text color, mirroring iOS `contrastingText` (threshold 0.5). */
private fun contrastingText(primary: Color): Color =
    if (primary.relativeLuminance() > 0.5f) Color.Black else Color.White

/** Resolve MLB team colors (stored as 0xRRGGBB longs) into a Compose [TeamColorPair]. */
fun mlbTeamColorPair(nameOrAbbrev: String): TeamColorPair {
    val c = MLBTeams.colors(nameOrAbbrev)
    return TeamColorPair(hexColor(c.primary), hexColor(c.secondary))
}

/**
 * MLB-specific formatting helpers — port of iOS `MLBFormatting`. Mirrors the
 * moneyline / run-line / total / date / game-time display used across the sheet.
 */
object MLBFormatting {
    fun moneyline(ml: Int?): String {
        if (ml == null) return "-"
        return if (ml > 0) "+$ml" else "$ml"
    }

    fun spread(s: Double?): String {
        if (s == null || s.isNaN()) return "-"
        val body = if (s == Math.rint(s)) s.toInt().toString() else String.format(Locale.US, "%.1f", s)
        return if (s > 0) "+$body" else body
    }

    fun line(value: Double?): String {
        if (value == null) return "-"
        return if (value == Math.rint(value)) value.toInt().toString()
        else String.format(Locale.US, "%.1f", value)
    }

    private val et: ZoneId = ZoneId.of("America/New_York")

    /** "EEE, MMM d" in ET from a `yyyy-MM-dd` official date. */
    fun dateLabel(raw: String): String = try {
        val date = java.time.LocalDate.parse(raw)
        date.format(DateTimeFormatter.ofPattern("EEE, MMM d", Locale.US))
    } catch (_: Exception) {
        raw
    }

    /** ISO UTC → "h:mm a ET"; "TBD" when missing/unparseable. */
    fun gameTime(raw: String?): String {
        if (raw.isNullOrEmpty()) return "TBD"
        return try {
            val instant = runCatching { ZonedDateTime.parse(raw).toInstant() }
                .getOrElse { java.time.Instant.parse(raw) }
            val zoned = instant.atZone(et)
            zoned.format(DateTimeFormatter.ofPattern("h:mm a", Locale.US)) + " ET"
        } catch (_: Exception) {
            "TBD"
        }
    }
}
